cuda/factorization/par_ilu_kernels.cu

```diff
@@ -60,9 +65,92 @@ constexpr int default_block_size{512};
 #include "common/factorization/par_ilu_kernels.hpp.inc"
 
 
+template <typename ValueType, typename IndexType>
+void add_diagonal_elements(std::shared_ptr<const DefaultExecutor> exec,
+                           matrix::Csr<ValueType, IndexType> *mtx,
+                           bool is_sorted)
+{
+    // TODO: Runtime can be optimized by choosing a appropriate size for the
+    //       subwarp dependent on the matrix properties
+    constexpr int subwarp_size = config::warp_size / 2;
+    auto mtx_size = mtx->get_size();
+    auto num_rows = static_cast<IndexType>(mtx_size[0]);
+    auto num_cols = static_cast<IndexType>(mtx_size[1]);
+    size_type row_ptrs_size = num_rows + 1;
+
+    Array<IndexType> row_ptrs_addition(exec, row_ptrs_size);
+    Array<bool> needs_change_host{exec->get_master(), 1};
+    needs_change_host.get_data()[0] = false;
+    Array<bool> needs_change_device{exec, 1};
+    needs_change_device = needs_change_host;
+
+    auto cuda_old_values = as_cuda_type(mtx->get_const_values());
+    auto cuda_old_col_idxs = as_cuda_type(mtx->get_const_col_idxs());
+    auto cuda_old_row_ptrs = as_cuda_type(mtx->get_row_ptrs());
+    auto cuda_row_ptrs_add = as_cuda_type(row_ptrs_addition.get_data());
+
+    const dim3 block_dim{default_block_size, 1, 1};
+    const dim3 grid_dim{
+        static_cast<uint32>(ceildiv(
+            num_rows, static_cast<size_type>(block_dim.x / subwarp_size))),
+        1, 1};
+    if (is_sorted) {
+        kernel::find_missing_diagonal_elements<true, subwarp_size>
+            <<<grid_dim, block_dim>>>(
+                num_rows, num_cols, cuda_old_col_idxs, cuda_old_row_ptrs,
+                cuda_row_ptrs_add,
+                as_cuda_type(needs_change_device.get_data()));
+    } else {
+        kernel::find_missing_diagonal_elements<false, subwarp_size>
+            <<<grid_dim, block_dim>>>(
+                num_rows, num_cols, cuda_old_col_idxs, cuda_old_row_ptrs,
+                cuda_row_ptrs_add,
+                as_cuda_type(needs_change_device.get_data()));
+    }
+    needs_change_host = needs_change_device;
+    if (!needs_change_host.get_const_data()[0]) {
+        return;
+    }
+
+    prefix_sum(exec, cuda_row_ptrs_add, row_ptrs_size);
+    exec->synchronize();
+
+    IndexType total_additions{};
+    exec->get_master()->copy_from(
+        exec.get(), 1, cuda_row_ptrs_add + row_ptrs_size - 1, &total_additions);
+    size_type new_num_elems = static_cast<size_type>(total_additions) +
+                              mtx->get_num_stored_elements();
+
+
+    Array<ValueType> new_values{exec, new_num_elems};
+    Array<IndexType> new_col_idxs{exec, new_num_elems};
+    auto cuda_new_values = as_cuda_type(new_values.get_data());
+    auto cuda_new_col_idxs = as_cuda_type(new_col_idxs.get_data());
```

Comment:
I guess you don't need the as_cuda_type here, since it is only necessary for complex

Reply:
That is true, however, I simply use `as_cuda_type` for all pointers I put into devices.